File: installing/env-installer/src/resolveOptionalSubdeps.ts
import util from 'node:util'

import { pickRegistryForPackage } from '@pnpm/config.pick-registry-for-package'
import { skippedOptionalDependencyLogger } from '@pnpm/core-loggers'
import { PnpmError } from '@pnpm/error'
import type { EnvLockfile } from '@pnpm/lockfile.fs'
import type { ResolvedDependencies } from '@pnpm/lockfile.types'
import { toLockfileResolution } from '@pnpm/lockfile.utils'
import type { createNpmResolver } from '@pnpm/resolving.npm-resolver'
import type { DependencyManifest, Registries } from '@pnpm/types'
import semver from 'semver'

type ResolveFromNpm = ReturnType<typeof createNpmResolver>['resolveFromNpm']

export interface ResolveOptionalSubdepsOpts {
  envLockfile: EnvLockfile
  lockfileDir: string
  registries: Registries
  resolveFromNpm: ResolveFromNpm
}

export async function resolveOptionalSubdeps (
  parentName: string,
  parentManifest: DependencyManifest,
  opts: ResolveOptionalSubdepsOpts
): Promise<ResolvedDependencies | undefined> {
  const optionalDeps = parentManifest.optionalDependencies
  if (!optionalDeps || Object.keys(optionalDeps).length === 0) {
    return undefined
  }

  const resolved: ResolvedDependencies = {}
  await Promise.all(Object.entries(optionalDeps).map(async ([subdepName, subdepSpec]) => {
    if (semver.valid(subdepSpec) == null) {
      // Ranges and tags would let the resolved version drift between machines
      // even with a stable parent integrity, breaking the lockfile's promise
      // of reproducible config-dep installs.
      throw new PnpmError(
        'CONFIG_DEP_OPTIONAL_NOT_EXACT',
        `Cannot install "${subdepName}@${subdepSpec}" as an optionalDependency of config dependency "${parentName}": only exact versions are supported (got "${subdepSpec}")`
      )
    }
    let resolution
    try {
      // `optional: true` opts into full registry metadata so the resolver
      // returns `libc` (and any other fields the abbreviated metadata strips).
      // See pnpm/pnpm#9950.
      resolution = await opts.resolveFromNpm({ alias: subdepName, bareSpecifier: subdepSpec, optional: true }, {
        lockfileDir: opts.lockfileDir,
        preferredVersions: {},
        projectDir: opts.lockfileDir,
      })
    } catch (err: unknown) {
      // Trust-downgrade is a security signal that must fail the install even
      // for optional deps; everything else mirrors npm's optionalDependencies
      // semantics — log and skip.
      if (util.types.isNativeError(err) && 'code' in err && err.code === 'ERR_PNPM_TRUST_DOWNGRADE') {
        throw err
      }
      skippedOptionalDependencyLogger.debug({
        details: util.types.isNativeError(err) ? err.toString() : String(err),
        package: {
          name: subdepName,
          // No resolved version yet; surface the requested specifier so log
          // consumers that format `${name}@${version}` don't render `@undefined`.
          version: subdepSpec,
          bareSpecifier: subdepSpec,
        },
        parents: [{ id: `${parentName}@${parentManifest.version}`, name: parentName, version: parentManifest.version }],
        prefix: opts.lockfileDir,
        reason: 'resolution_failure',
      })
      return
    }
    if (
      resolution?.resolution == null ||
      !('integrity' in resolution.resolution) ||
      typeof resolution.resolution.integrity !== 'string' ||
      !resolution.resolution.integrity ||
      resolution.manifest == null
    ) {
      throw new PnpmError(
        'BAD_CONFIG_DEP',
        `Cannot resolve optionalDependency "${subdepName}" of config dependency "${parentName}" because it has no integrity`
      )
    }
    const subdepVersion = resolution.manifest.version
    const registry = pickRegistryForPackage(opts.registries, subdepName)
    const subdepKey = `${subdepName}@${subdepVersion}`

    opts.envLockfile.packages[subdepKey] = {
      resolution: toLockfileResolution(
        { name: subdepName, version: subdepVersion },
        resolution.resolution,
        registry
      ),
      ...pickPlatformFields(resolution.manifest),
    }
    if (opts.envLockfile.snapshots[subdepKey] == null) {
      opts.envLockfile.snapshots[subdepKey] = { optional: true }
    }
    resolved[subdepName] = subdepVersion
  }))

  return Object.keys(resolved).length > 0 ? resolved : undefined
}

function pickPlatformFields (manifest: DependencyManifest): { os?: string[], cpu?: string[], libc?: string[] } {
  const out: { os?: string[], cpu?: string[], libc?: string[] } = {}
  if (manifest.os?.length) out.os = manifest.os
  if (manifest.cpu?.length) out.cpu = manifest.cpu
  if (manifest.libc?.length) out.libc = manifest.libc
  return out
}

File: pacquet/crates/env-installer/src/resolve_optional_subdeps.rs
//! Resolve one level of a config dependency's `optionalDependencies`
//! into the env lockfile. Mirrors pnpm's
//! [`resolveOptionalSubdeps`](https://github.com/pnpm/pnpm/blob/31858c544b/installing/env-installer/src/resolveOptionalSubdeps.ts).
//!
//! Only exact versions are accepted: a range or tag would let the
//! resolved version drift between machines even with a stable parent
//! integrity, breaking the lockfile's reproducibility promise.

use crate::{
    ConfigDepError, manifest_lockfile::package_metadata, options::ConfigDepsInstallOptions,
};
use pacquet_lockfile::{
    EnvLockfile, PackageKey, PkgName, PkgVerPeer, SnapshotDepRef, SnapshotEntry,
};
use pacquet_resolving_resolver_base::{ResolveOptions, Resolver, WantedDependency};
use std::collections::HashMap;

/// Resolve `parent_manifest.optionalDependencies` and record each into
/// `env_lockfile`'s `packages` + `snapshots`. Returns the
/// `optionalDependencies` map (alias → version) for the parent's
/// snapshot, or `None` when the parent declares none.
pub async fn resolve_optional_subdeps(
    parent_name: &str,
    parent_manifest: &serde_json::Value,
    resolver: &dyn Resolver,
    opts: &ConfigDepsInstallOptions<'_>,
    env_lockfile: &mut EnvLockfile,
) -> Result<Option<HashMap<PkgName, SnapshotDepRef>>, ConfigDepError> {
    let Some(optional_deps) =
        parent_manifest.get("optionalDependencies").and_then(|value| value.as_object())
    else {
        return Ok(None);
    };
    if optional_deps.is_empty() {
        return Ok(None);
    }

    let mut resolved: HashMap<PkgName, SnapshotDepRef> = HashMap::new();
    for (subdep_name, subdep_spec) in optional_deps {
        let subdep_spec = subdep_spec.as_str().unwrap_or_default();
        if subdep_spec.parse::<node_semver::Version>().is_err() {
            return Err(ConfigDepError::OptionalNotExact {
                parent_name: parent_name.to_string(),
                subdep_name: subdep_name.clone(),
                spec: subdep_spec.to_string(),
            });
        }

        let wanted = WantedDependency {
            alias: Some(subdep_name.clone()),
            bare_specifier: Some(subdep_spec.to_string()),
            optional: Some(true),
            ..WantedDependency::default()
        };
        let resolve_opts = ResolveOptions {
            project_dir: opts.root_dir.to_path_buf(),
            lockfile_dir: opts.root_dir.to_path_buf(),
            ..ResolveOptions::default()
        };
        let result = resolver
            .resolve(&wanted, &resolve_opts)
            .await
            .map_err(|error| ConfigDepError::Resolve {
                spec: format!("{subdep_name}@{subdep_spec}"),
                error,
            })?
            .ok_or_else(|| ConfigDepError::BadConfigDep {
                message: format!(
                    r#"Cannot resolve optionalDependency "{subdep_name}" of config dependency "{parent_name}" because it has no integrity"#,
                ),
            })?;

        let Some(name_ver) = result.name_ver.as_ref() else {
            return Err(ConfigDepError::BadConfigDep {
                message: format!(
                    r#"Cannot resolve optionalDependency "{subdep_name}" of config dependency "{parent_name}" because it has no integrity"#,
                ),
            });
        };
        if !resolution_has_integrity(&result.resolution) {
            return Err(ConfigDepError::BadConfigDep {
                message: format!(
                    r#"Cannot resolve optionalDependency "{subdep_name}" of config dependency "{parent_name}" because it has no integrity"#,
                ),
            });
        }
        let subdep_version = name_ver.suffix.to_string();
        let registry = opts.pick_registry(subdep_name);
        let pkg_key: PackageKey = format!("{subdep_name}@{subdep_version}")
            .parse()
            .map_err(|_| ConfigDepError::BadConfigDep {
                message: format!("Resolved optionalDependency {subdep_name}@{subdep_version} has an unparsable key"),
            })?;

        env_lockfile.packages.insert(
            pkg_key.clone(),
            package_metadata(subdep_name, &subdep_version, &result, registry, false),
        );
        env_lockfile
            .snapshots
            .entry(pkg_key)
            .or_insert_with(|| SnapshotEntry { optional: true, ..SnapshotEntry::default() });

        let ver_peer =
            subdep_version.parse::<PkgVerPeer>().map_err(|_| ConfigDepError::BadConfigDep {
                message: format!(
                    "Resolved optionalDependency version {subdep_version} is not a valid version",
                ),
            })?;
        let pkg_name: PkgName = subdep_name.parse().map_err(|_| ConfigDepError::BadConfigDep {
            message: format!("Resolved optionalDependency name {subdep_name} is invalid"),
        })?;
        resolved.insert(pkg_name, SnapshotDepRef::Plain(ver_peer));
    }

    Ok((!resolved.is_empty()).then_some(resolved))
}

pub(crate) fn resolution_has_integrity(resolution: &pacquet_lockfile::LockfileResolution) -> bool {
    use pacquet_lockfile::LockfileResolution;
    match resolution {
        LockfileResolution::Registry(_) => true,
        LockfileResolution::Tarball(tarball) => tarball.integrity.is_some(),
        _ => false,
    }
}
